### ISSUE TITLE
Add additional details on Gradle Check failures autocut issues

### ISSUE THREAD

**Issue by prudhvigodithi** [opensearch-project/OpenSearch#13950]
### Is your feature request related to a problem? Please describe

Coming from https://github.com/opensearch-project/OpenSearch/issues/11217 and https://github.com/opensearch-project/OpenSearch/issues/3713 we now have the the failure analytics for OpenSearch Gradle Check failures. For more details check the [Developer Guide](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#gradle-check-metrics-dashboard). The next step is to highlight the flaky failures as GitHub Issues so they can be prioritized and fixed. Additionally, when a test fails during the creation of a new PR but is not related to the PR's code changes, this process will help make the contributor aware of the existing flaky test.

### Describe the solution you'd like

From the [OpenSearch Gradle Check Metrics](https://metrics.opensearch.org/_dashboards/app/dashboards#/view/e5e64d40-ed31-11ee-be99-69d1dbc75083)[OpenSearch Gradle Check Metrics](https://metrics.opensearch.org/_dashboards/app/dashboards#/view/e5e64d40-ed31-11ee-be99-69d1dbc75083) dashboard create an issue (with specific format) on failing tests that are part of post merge actions. The failing tests from post merge actions which are executed after the PR is merged are for sure the flaky tests.
<img width="1433" alt="Screenshot 2024-06-03 at 11 50 59 AM" src="https://github.com/opensearch-project/OpenSearch/assets/28733332/aa046374-183b-4d78-93df-fe4936bc8dd2">

### Related component

Build

### Describe alternatives you've considered

Today the gradle check failure issues are created from https://github.com/opensearch-project/OpenSearch/blob/main/.github/workflows/gradle-check.yml#L161-L168 which sometimes fails to execute https://github.com/opensearch-project/OpenSearch/actions/runs/9320653340/job/25657907035, so clean up the issues and disable the functionality in the gradle-check.yml.

### Additional context

Coming from @dblock https://github.com/opensearch-project/OpenSearch/issues/11217#issuecomment-2145772021 we should also 

1. For any failed test, lookup an existing flaky test issue, if it exists, comment.
2. For any new failure, highlight it in comments with something like "new flaky test? please check and open one manually".

**Comment by msfroh** [comment]
I added a comment on https://github.com/opensearch-project/OpenSearch/issues/11217#issuecomment-2146250437, highlighting that we need to avoid tainted data from tests broken by changes in an open PR, but that get fixed before the PR is merged. (Short story: I added a commit to fix 1 test and broke over 1000 other tests. Nobody else saw it though, because my mistake was limited to my open PR.)

Maybe a good heuristic could look at tests that fail across multiple PRs in some time window? Alternatively, if we have a job that just runs Gradle check continuously (without changing any code), it could be a good canary to collect "true" failures.

**Comment by prudhvigodithi** [comment]
> Short story: I added a commit to fix 1 test and broke over 1000 other tests. Nobody else saw it though, because my mistake was limited to my open PR.

Thanks @msfroh, the question I have is for `I added a commit to fix 1 test and broke over 1000 other tests` are you referring after the PR was merged and it broke 1000 other tests? AFIAK it should be caught while the PR is open itself because today we run gradle check against the commit part of the open PR and against the commit that was merged (post merge action).

**Comment by msfroh** [comment]
> are you referring after the PR was merged and it broke 1000 other tests? AFIAK it should be caught while the PR is open itself because today we run gradle check against the commit part of the open PR 

Correct -- it was caught while the PR was still open. I noticed that those 1000+ test failures showed up on the dashboard (and made it look like `ClientYamlTestSuiteIT` is the biggest source of failures in the past week).

**Comment by peternied** [comment]
[Triage - attendees [1](https://github.com/srikanthpadakanti) [2](https://github.com/peternied) [3](https://github.com/lukas-vlcek) [4](https://github.com/jed326) [5](https://github.com/reta) [6](https://github.com/andrross) [7](https://github.com/kotwanikunal)]
@prudhvigodithi Thanks for creating this issue

**Comment by prudhvigodithi** [comment]
> Maybe a good heuristic could look at tests that fail across multiple PRs in some time window? Alternatively, if we have a job that just runs Gradle check continuously (without changing any code), it could be a good canary to collect "true" failures.

Hey @msfroh thanks for your input, the idea is to create a issue on when a test is failed in Post Merge Action (after the PR is merged) not for the failed tests that are part of the open PR, so in your case the tests that are caught while the PR is open, the automaton will not create issues for these. Once the PR is merged by the maintainers after the gradle check is green, there is one more gradle check that gets triggered based on the merge commit, so whatever tests fail as part of this commit are considered flaky as they just showed green for the PR to get merged. The metrics project is collecting this data and will use it for issue creation.
<img width="1421" alt="Screenshot 2024-06-05 at 10 00 42 AM" src="https://github.com/opensearch-project/OpenSearch/assets/28733332/4bc7e472-7158-4c2b-ac81-cd845b327f3b">
Thank you

**Comment by prudhvigodithi** [comment]
Hey @reta coming from https://github.com/opensearch-project/OpenSearch/issues/14012 I can see the test `org.opensearch.cache.store.disk.EhCacheDiskCacheTests.testComputeIfAbsentConcurrently` failure here in the [Metrics Dashboard](https://metrics.opensearch.org/_dashboards/app/dashboards?security_tenant=global#/view/e5e64d40-ed31-11ee-be99-69d1dbc75083?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:now-90d,to:now))&_a=(description:'',filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'2e729590-eb26-11ee-b008-6583d3dd450c',key:test_name,negate:!f,params:(query:org.opensearch.cache.store.disk.EhCacheDiskCacheTests.testComputeIfAbsentConcurrently),type:phrase),query:(match_phrase:(test_name:org.opensearch.cache.store.disk.EhCacheDiskCacheTests.testComputeIfAbsentConcurrently)))),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!t,title:'OpenSearch%20Gradle%20Check%20Report',viewMode:view)), I dont see this test failing as part of the Post Merge Actions, but seen on failing on an open PR https://github.com/opensearch-project/OpenSearch/pull/13772, is there way to better mechanism to identify the Flaky tests apart from that are failing from Post Merge Actions? or its based on the judgement of PR creator when the modified code and failure tests are not corelated.
Thanks

**Comment by reta** [comment]
> or its based on the judgement of PR creator when the modified code and failure tests are not corelated.

Hey @prudhvigodithi , this is fair point, I will close the https://github.com/opensearch-project/OpenSearch/issues/14012 for now since indeed the code is not merged but should not be in any correlation. Thanks for bringing it up!

**Comment by prudhvigodithi** [comment]
Thanks @reta, but there could be a chance where it was lucky and have not failed in post merge actions :) we should even have a mechanism to flag these type of flaky issues as well that are part of the open PR.

**Comment by prudhvigodithi** [comment]
I will start with an automation by creating issues at class level in the following format. I have noticed one class can have multiple failing tests so rather than creating multiple issues for each failing test we can group at class level. For example the class `IndicesRequestCacheIT` has multiple failing tests for different PR's. The automation upon scheduled cron will auto refresh the list updating the issue body. The GitHub Issue will be created as follows (the Github UI will take care of linking the PR's with the issue once the issue mentions the PR):



### Title: 
[AUTOCUT] The Gradle check encountered flaky failures with `IndicesRequestCacheIT`.

Noticed the `IndicesRequestCacheIT` has some flaky, failing tests that failed during **post-merge actions**.

| Git Reference                                                                | Merged Pull Request                                                                 | Build Details                                                                 | Test Name                                                                                                                                                       |
|------------------------------------------------------------------------------|------------------------------------------------------------------------------|--------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| e9b6a8d652d51e26fe10159d0ebf4c8b429b5c28                                      | [13590](https://github.com/opensearch-project/OpenSearch/pull/13590)        | [40040](https://build.ci.opensearch.org/job/40040/testReport/)                | `org.opensearch.indices.IndicesRequestCacheIT.testDynamicStalenessThresholdUpdate {p0={"opensearch.experimental.feature.pluggable.caching.enabled":"true"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode {p0={"search.concurrent_segment_search.enabled":"true"}}` |
| de13acab68cb9b336e167accce457afea04c92a1                                      | [13920](https://github.com/opensearch-project/OpenSearch/pull/13920)        | [39716](https://build.ci.opensearch.org/job/39716/testReport/)                | `org.opensearch.indices.IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode {p0={"search.concurrent_segment_search.enabled":"true"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode {p0={"search.concurrent_segment_search.enabled":"false"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode {p0={"opensearch.experimental.feature.pluggable.caching.enabled":"true"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode {p0={"opensearch.experimental.feature.pluggable.caching.enabled":"false"}}` |
| b06d0b93b7565c577e6e141967defac3dd888696                                      | [14062](https://github.com/opensearch-project/OpenSearch/pull/14062)        | [40165](https://build.ci.opensearch.org/job/40165/testReport/)                | `org.opensearch.indices.IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode {p0={"search.concurrent_segment_search.enabled":"true"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode {p0={"opensearch.experimental.feature.pluggable.caching.enabled":"true"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.testDeleteAndCreateSameIndexShardOnSameNode {p0={"opensearch.experimental.feature.pluggable.caching.enabled":"false"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.testCacheCleanupWithDefaultSettings {p0={"search.concurrent_segment_search.enabled":"false"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.testCacheCleanupOnEqualStalenessAndThreshold {p0={"search.concurrent_segment_search.enabled":"true"}}`<br><br>`org.opensearch.indices.IndicesRequestCacheIT.classMethod` |


The other pull requests, besides those involved in post-merge actions, that contain failing tests with the `IndicesRequestCacheIT` class are:

- [13865](https://github.com/opensearch-project/OpenSearch/pull/13865)
- [13172](https://github.com/opensearch-project/OpenSearch/pull/13172)
- [13801](https://github.com/opensearch-project/OpenSearch/pull/13801)
- [13934](https://github.com/opensearch-project/OpenSearch/pull/13934)
- [13997](https://github.com/opensearch-project/OpenSearch/pull/13997)
- [12982](https://github.com/opensearch-project/OpenSearch/pull/12982)
- [13590](https://github.com/opensearch-project/OpenSearch/pull/13590)
- [13998](https://github.com/opensearch-project/OpenSearch/pull/13998)
- [13374](https://github.com/opensearch-project/OpenSearch/pull/13374)
- [13988](https://github.com/opensearch-project/OpenSearch/pull/13988)
- [13577](https://github.com/opensearch-project/OpenSearch/pull/13577)
- [13821](https://github.com/opensearch-project/OpenSearch/pull/13821)
- [14026](https://github.com/opensearch-project/OpenSearch/pull/14026)
- [13888](https://github.com/opensearch-project/OpenSearch/pull/13888)

For more details on the failed tests refer to [OpenSearch Gradle Check Metrics](https://metrics.opensearch.org/_dashboards/app/dashboards#/view/e5e64d40-ed31-11ee-be99-69d1dbc75083) dashboard.


@getsaurabh02 @dblock @reta @msfroh @andrross

**Comment by reta** [comment]
> The GitHub Issue will be created as follows (the Github UI will take care of linking the PR's with the issue once the issue mentions the PR):

Thanks @prudhvigodithi , it looks awesome to start with

**Comment by prudhvigodithi** [comment]
Hey Just an update on this, I have the issues created in my fork repo with automation
https://github.com/prudhvigodithi/OpenSearch/issues/14 
https://github.com/prudhvigodithi/OpenSearch/issues/14 
I have the PR open https://github.com/opensearch-project/opensearch-build-libraries/pull/436 and under review for pushing the library that can run this automation form our [Jenkins](https://build.ci.opensearch.org/) querying the [metrics cluster](https://metrics.opensearch.org/).

Also what I noticed is the label used today is `>test-failure` coming from the [template](https://github.com/opensearch-project/OpenSearch/blob/main/.github/ISSUE_TEMPLATE/failed_check.md?plain=1#L3), just curious if  `>` is an error or was purposely added? @andrross @dblock @reta @peternied ?
Moving forward we can use something like `test-failure`.
Thank you 
@getsaurabh02

**Comment by reta** [comment]
> Also what I noticed is the label used today is `>test-failure` coming from the [template](https://github.com/opensearch-project/OpenSearch/blob/main/.github/ISSUE_TEMPLATE/failed_check.md?plain=1#L3), just curious if `>` is an error or was purposely added? @andrross @dblock @reta @peternied ?

I know we use `>` but have not context why (my guess is to highlight such labels as  more important than others)

**Comment by prudhvigodithi** [comment]
Thanks @reta and @andrross, in that case ya we can continue to use `>test-failure`, during issue creation with automation I will as well add the `>test-failure` label.

**Comment by prudhvigodithi** [comment]
The automation flagged and created the following issues which are identifying as flaky tests from past one month.

https://github.com/opensearch-project/OpenSearch/issues/14332
https://github.com/opensearch-project/OpenSearch/issues/14331
https://github.com/opensearch-project/OpenSearch/issues/14330
https://github.com/opensearch-project/OpenSearch/issues/14330
https://github.com/opensearch-project/OpenSearch/issues/14328
https://github.com/opensearch-project/OpenSearch/issues/14327
https://github.com/opensearch-project/OpenSearch/issues/14326
https://github.com/opensearch-project/OpenSearch/issues/14325
https://github.com/opensearch-project/OpenSearch/issues/14324
https://github.com/opensearch-project/OpenSearch/issues/14323
https://github.com/opensearch-project/OpenSearch/issues/14322
https://github.com/opensearch-project/OpenSearch/issues/14321
https://github.com/opensearch-project/OpenSearch/issues/14320
https://github.com/opensearch-project/OpenSearch/issues/14319
https://github.com/opensearch-project/OpenSearch/issues/14318
https://github.com/opensearch-project/OpenSearch/issues/14317
https://github.com/opensearch-project/OpenSearch/issues/14316
https://github.com/opensearch-project/OpenSearch/issues/14315
https://github.com/opensearch-project/OpenSearch/issues/14314
https://github.com/opensearch-project/OpenSearch/issues/14313
https://github.com/opensearch-project/OpenSearch/issues/14312
https://github.com/opensearch-project/OpenSearch/issues/14311
https://github.com/opensearch-project/OpenSearch/issues/14310
https://github.com/opensearch-project/OpenSearch/issues/14309
https://github.com/opensearch-project/OpenSearch/issues/14308
https://github.com/opensearch-project/OpenSearch/issues/14307
https://github.com/opensearch-project/OpenSearch/issues/14306
https://github.com/opensearch-project/OpenSearch/issues/14305
https://github.com/opensearch-project/OpenSearch/issues/14304
https://github.com/opensearch-project/OpenSearch/issues/14303
https://github.com/opensearch-project/OpenSearch/issues/14302
https://github.com/opensearch-project/OpenSearch/issues/14301
https://github.com/opensearch-project/OpenSearch/issues/14300
https://github.com/opensearch-project/OpenSearch/issues/14299
https://github.com/opensearch-project/OpenSearch/issues/14298
https://github.com/opensearch-project/OpenSearch/issues/14297
https://github.com/opensearch-project/OpenSearch/issues/14296
https://github.com/opensearch-project/OpenSearch/issues/14295
https://github.com/opensearch-project/OpenSearch/issues/14294
https://github.com/opensearch-project/OpenSearch/issues/14293
https://github.com/opensearch-project/OpenSearch/issues/14292
https://github.com/opensearch-project/OpenSearch/issues/14291
https://github.com/opensearch-project/OpenSearch/issues/14290
https://github.com/opensearch-project/OpenSearch/issues/14289
https://github.com/opensearch-project/OpenSearch/issues/14288
https://github.com/opensearch-project/OpenSearch/issues/14287

Adding @andrross @reta @dblock @msfroh @getsaurabh02 to please take a look.

**Comment by prudhvigodithi** [comment]
I just created a PR https://github.com/opensearch-project/OpenSearch/pull/14334 to remove the issue creation from the gradle check workflow.

**Comment by prudhvigodithi** [comment]
The 2nd time the automation wont create new issues but rather updates the existing issue body if an issue for a flaky test already exists. FYI Re-ran the job to validate this https://build.ci.opensearch.org/job/gradle-check-flaky-test-detector/4/console.

**Comment by prudhvigodithi** [comment]
Hey FYI, I have added a new visualizations to [OpenSearch Gradle Check Metrics](https://metrics.opensearch.org/_dashboards/app/dashboards#/view/e5e64d40-ed31-11ee-be99-69d1dbc75083) dashboard to track the trend of these flaky issues.

![Screenshot 2024-06-13 at 3 58 49 PM](https://github.com/opensearch-project/OpenSearch/assets/28733332/8713778c-0b06-4688-aae0-5f5af9675dd9)

**Comment by reta** [comment]
> Adding @andrross @reta @dblock @msfroh @getsaurabh02 to please take a look.

It looks pretty cool, thanks @prudhvigodithi ! Just one question, I noticed the `flaky-test` label is not there (but there is a new one `> test-failure`, that's on purpose? Thanks!

**Comment by dblock** [comment]
Should we find a way to link/close the existing manually created flaky test issues and assign the automatically cut ones to the same devs?

**Comment by prudhvigodithi** [comment]
> It looks pretty cool, thanks @prudhvigodithi ! Just one question, I noticed the `flaky-test` label is not there (but there is a new one `> test-failure`, that's on purpose? Thanks!

Hey @reta I just referred this issue in past https://github.com/opensearch-project/OpenSearch/issues/14255 and added `>test-failure` label, I can also add the `flaky-test` to the automation and it will update the existing open issues as well. Please let me know. Do we need both `flaky-test` and `>test-failure` ?

**Comment by prudhvigodithi** [comment]
> Should we find a way to link/close the existing manually created flaky test issues and assign the automatically cut ones to the same devs?

Sure @dblock, we should do that.

**Comment by reta** [comment]
> Please let me know. Do we need both `flaky-test` and `>test-failure` ?

That would help to preserve any existing boards / filters, thanks @prudhvigodithi

**Comment by prudhvigodithi** [comment]
Can I get the following backport PR's approved please?
https://github.com/opensearch-project/OpenSearch/pull/14352
https://github.com/opensearch-project/OpenSearch/pull/14351
https://github.com/opensearch-project/OpenSearch/pull/14350
https://github.com/opensearch-project/OpenSearch/pull/14349

**Comment by prudhvigodithi** [comment]
Thanks @dblock.

**Comment by prudhvigodithi** [comment]
Hey @andrross @dblock @reta, since we have the automation in place please let me know if we can close this issue ?

For: 
> Should we find a way to link/close the existing manually created flaky test issues and assign the automatically cut ones to the same devs?

There are [145 manual created issue](https://github.com/opensearch-project/OpenSearch/issues?q=is%3Aissue+is%3Aopen+label%3Aflaky-test+-author%3Aopensearch-ci-bot) for the flaky tests, we should go over them and close if there is already an automation issue created for the same flaky test.

@getsaurabh02

**Comment by reta** [comment]
> Hey @andrross @dblock @reta, since we have the automation in place please let me know if we can close this issue ?

I think we could close this issue indeed, thanks @prudhvigodithi

**Comment by andrross** [comment]
Closing. Thanks @prudhvigodithi!
